### PR TITLE
Stager first ping from server and suppress pings if a ping was received.

### PR DESCRIPTION
**Pull request by derekcollison**
Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -2635,10 +2640,15 @@ func (c *client) processPingTimer() {
...
-	if delta := time.Since(c.last); delta < c.srv.getOpts().PingInterval {
-		c.Debugf("Delaying PING due to activity %v ago", delta.Round(time.Second))
+	// If we have had activity within the PingInterval then
+	// there is no need to send a ping. This can be client data
+	// or if we received a ping from the other side.
+	pingInterval := c.srv.getOpts().PingInterval
+
+	if delta := time.Since(c.last); delta < pingInterval {
+		c.Debugf("Delaying PING due to client activity %v ago", delta.Round(time.Second))
+	} else if delta := time.Since(c.ping.last); delta < pingInterval {
```
Comment (kozlovic) [inline server/client.go]: The problem I see here is that we use PING/PONG to compute/update the connection's RTT (I know I use that for gateways). With this change, I am not sure a connection RTT would be ever be computed (except in monitor/events).

File: server/client.go [hunk 2]
```diff
@@ -1461,6 +1462,10 @@ func (c *client) processPing() {
 	}
 	c.sendPong()
 
+	// Record this to suppress us sending one if this
+	// is withing a given time interval for activity.
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/client.go]: will fix.

File: server/client.go [hunk 3]
```diff
@@ -2655,6 +2665,20 @@ func (c *client) processPingTimer() {
 	c.setPingTimer()
 }
 
+// Lock should be held
+// We Randomize the first one by an offset up to 20%, e.g. 2m ~= max 24s.
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/client.go]: will fix

File: test/ping_test.go [hunk 4]
```diff
@@ -187,3 +187,42 @@ func TestUnpromptedPong(t *testing.T) {
...
+		t.Fatalf("pingTime too high: %v", pingTime)
+	}
+
+	time.Sleep(pingInterval / 2)
+
+	// Sending a PING should suppress.
+	send("PING\r\n")
+	expect(pongRe)
+
+	// This will wait for
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply test/ping_test.go]: Will fix

File: test/ping_test.go [hunk 5]
```diff
@@ -187,3 +187,42 @@ func TestUnpromptedPong(t *testing.T) {
...
+	expectNothingTimeout(t, c, time.Now().Add(100*time.Millisecond))
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply test/ping_test.go]: No I wanted it explicit here to match the PingInterval.

File: server/client.go (hunk 1 above)
Comment (derekcollison) [reply server/client.go]: Hmm, that is a good point. Let me think about this one for a bit.

File: server/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/client.go]: ok take a look now.

File: server/client.go (hunk 1 above)
Comment (derekcollison) [reply server/client.go]: ok take a look now.

File: server/const.go [hunk 6]
```diff
@@ -156,4 +156,8 @@ const (
 	// attempt to reconnect a route, gateway or leaf node connection.
 	// The default is to report every attempt.
 	DEFAULT_RECONNECT_ERROR_REPORTS = 1
+
+	// DEFAULT_RTT_MEASUREMENT_INTERVAL	is how often we want to measure RTT from
```
Comment (kozlovic) [inline server/const.go]: check spaces after const name...

**Pull request merged**